### PR TITLE
Create dealii::dealii target and add compile options

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request makes it possible to use
```
target_link_libraries(${TARGET} dealii::dealii)
```
including using `DEAL_II_CXX_FLAGS`.

(tamiko:) In reference to #14947

### CODE REVIEW

**Comment by sebproell** [comment]
@masterleinad I now have `DEBUG` defined in my release build when linking to `dealii::dealii` :D


When I look into the installed deal.IITargets.cmake, I see

```
set_target_properties(dealii::dealii PROPERTIES
  INTERFACE_LINK_LIBRARIES "dealii::dealii_debug;dealii::dealii_release"
)
```

which doesn't look quite right.

**Comment by masterleinad** [comment]
@sebproell With the latest commit, we know do
```CMake
# Create imported target dealii::dealii
add_library(dealii::dealii INTERFACE IMPORTED)
  
set_target_properties(dealii::dealii PROPERTIES
  INTERFACE_LINK_LIBRARIES "\$<\$<CONFIG:Debug>:dealii::dealii_debug>;\$<\$<CONFIG:>:dealii::dealii_debug>;\$<\$<CONFIG:Release>:dealii::dealii_release>"
) 
```

**Comment by tamiko** [comment]
@masterleinad I am currently running the FE Rodeo. Would you mind to give me 48h to think about this approach.

In general I think it is neat to introduce a `dealii::dealii` dummy target but then again we have the problem to hardcode it to `Debug` and `Release`.

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -127,6 +130,14 @@ foreach(build ${DEAL_II_BUILD_TYPES})
     RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${DEAL_II_EXECUTABLE_RELDIR}"
     )
 
+  separate_arguments(_compile_options UNIX_COMMAND
+    "${DEAL_II_CXX_FLAGS} ${DEAL_II_CXX_FLAGS_${build_uppercase}}"
+    )
+  shell_escape_option_groups(_compile_options)
+  target_compile_options(${DEAL_II_NAMESPACE}_${build_lowercase} PUBLIC
+    $<$<COMPILE_LANGUAGE:CXX>:${_compile_options}>
+    )
```
Comment (tamiko) [inline source/CMakeLists.txt:139]: This is a big issue:

As neat as it seems we have to be very careful with adding `PUBLIC` compile and link options to the targets. 
Because this cannot be overriden by user code any more.

What I mean is this:

Suppose deal.II had been configured with `-std=c++17` (and the compiler is defaulting to `-std=c++20`. Then neither
```
set(CMAKE_CXX_FLAGS "-std=c++20")
add_executable(${TARGET} ${TARGET_SRC})
target_link_libraries(${TARGET} dealii::dealii)
```
nor
```
add_executable(${TARGET} ${TARGET_SRC})
set_property(TARGET ${TARGET} PROPERTY COMPILE_OPTIONS "-std=c++20")
target_link_libraries(${TARGET} dealii::dealii)
```
will allow you to override this compile flag any more. CMake will always order the compile options that come with `dealii::dealii_(debug|release)` last.


Remark: Funny enough, `set_property(TARGET ${TARGET} PROPERTY CXX_STANDARD 20)` works because in this case the language switch will be applied last... but this isnot the point I want to make here. Think for example about packaging for Debian where we are forced to have a `-march=generic` but suppose you want to use `-march=native` for your own target. This will not work.

The bottom line is that I think that the `dealii::dealii_debug` and `dealii::dealii_release` target will have to be free from any compile and link options.

So what about adding these target properties to the `dealii::dealii` dummy target instead? (with the same `$<CONFIG:...>` generator expression logic).

**Review by tamiko — CHANGES_REQUESTED** [review]
I would like to get some more refactoring in before we make the design decision to include a `dealii::dealii` target. I am not opposed to this change - to the contrary, I think it is a neat idea.

**Comment by tamiko** [comment]
@sebproell For the time being `target_link_libraries(TARGET ${DEAL_II_TARGET})` should just work. It uses the inelegant `optimized` and `debug` "keywords", though...

File: examples/step-1/CMakeLists.txt [hunk 2]
```diff
@@ -34,6 +35,5 @@ if(NOT ${deal.II_FOUND})
     )
 endif()
 
-deal_ii_initialize_cached_variables()
-project(${TARGET})
-deal_ii_invoke_autopilot()
+add_executable(${TARGET} ${TARGET_SRC})
+target_link_libraries(${TARGET} dealii::dealii)
```
Comment (tamiko) [inline examples/step-1/CMakeLists.txt:39]: Would you mind to drop all changes to the example steps for the time being?
I would like to address this in a separate pull request.

In particular, I am not happy with this change:

 I think it is important that we still support the quirky setup in our example step. Your change will break our compiler autoselection and will remove the `debug`, `release`, `run` and cleanup targets. Therefore, I really think that we should keep the `deal_ii_invoke_autopilot()` macro around.

File: source/CMakeLists.txt [hunk 3]
```diff
@@ -127,6 +142,24 @@ foreach(build ${DEAL_II_BUILD_TYPES})
     RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${DEAL_II_EXECUTABLE_RELDIR}"
     )
 
+  separate_arguments(_compile_options UNIX_COMMAND
+    "${DEAL_II_CXX_FLAGS_${build_uppercase}}"
```
Comment (tamiko) [inline source/CMakeLists.txt:146]:
```suggestion
    "${DEAL_II_CXX_FLAGS_${build}}"
```

File: source/CMakeLists.txt [hunk 4]
```diff
@@ -170,8 +203,31 @@ foreach(build ${DEAL_II_BUILD_TYPES})
...
     ARCHIVE DESTINATION ${DEAL_II_LIBRARY_RELDIR}
     )
+
+  if(${build_lowercase} MATCHES "debug")
+    target_link_libraries(${DEAL_II_TARGET_NAME} INTERFACE "$<$<CONFIG:Debug>:${DEAL_II_TARGET_NAME}::${DEAL_II_TARGET_NAME}_${build_lowercase}>")
+    target_link_libraries(${DEAL_II_TARGET_NAME} INTERFACE "$<$<CONFIG:>:${DEAL_II_TARGET_NAME}::${DEAL_II_TARGET_NAME}_${build_lowercase}>")
+  endif()
+  if(${build_lowercase} MATCHES "release")
+    target_link_libraries(${DEAL_II_TARGET_NAME} INTERFACE "$<$<CONFIG:Release>:${DEAL_II_TARGET_NAME}::${DEAL_II_TARGET_NAME}_${build_lowercase}>")
+  endif()
```
Comment (tamiko) [inline source/CMakeLists.txt:213]:
```suggestion
  target_link_libraries(${DEAL_II_TARGET_NAME} INTERFACE "$<$<CONFIG:${build_camelcase}>:${DEAL_II_TARGET_NAME}::${DEAL_II_TARGET_NAME}_${build_lowercase}>")
```

Same here.

**Review by tamiko — CHANGES_REQUESTED** [review]
Looks good, I have some minor technical suggestions.

File: source/CMakeLists.txt [hunk 5]
```diff
@@ -75,8 +75,23 @@ add_subdirectory(sundials)
...
+separate_arguments(_compile_options UNIX_COMMAND "${DEAL_II_CXX_FLAGS}")
+shell_escape_option_groups(_compile_options)
+target_compile_options(${DEAL_II_TARGET_NAME} INTERFACE
+  $<$<COMPILE_LANGUAGE:CXX>:${_compile_options}>
+  )
+
+
 foreach(build ${DEAL_II_BUILD_TYPES})
   string(TOLOWER ${build} build_lowercase)
+  string(TOUPPER ${build} build_uppercase)
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply source/CMakeLists.txt]: Hmm... I'm pretty sure I needed that initially but I can confirm that `${build}` is uppercase indeed.

**Comment by tamiko** [comment]
@masterleinad Would you mind to squash the reverts?

**Comment by masterleinad** [comment]
> @masterleinad Would you mind to squash the reverts?

I just squashed everything.
